src/wp-includes/class-wp-tax-query.php

```diff
@@ -633,10 +633,15 @@ public function transform_query( &$query, $resulting_field ) {
 				$args['term_taxonomy_id'] = $terms;
 				break;
 			default:
-				$args['include'] = wp_parse_id_list( $terms );
+				$terms           = wp_parse_id_list( $terms );
+				$args['include'] = $terms;
 				break;
 		}
 
+		if ( ! is_taxonomy_hierarchical( $query['taxonomy'] ) ) {
+			$args['number'] = count( $terms );
+		}
```

Comment:
Based on the comments above, how about we still rely on `$args['include']`, which makes this a bit safer?
```suggestion
		if ( isset( $args['include'] ) && ! is_taxonomy_hierarchical( $query['taxonomy'] ) ) {
			$args['number'] = count( $args['include'] );
		}
```

Reply:
But what about query where `term_taxonomy_id` are set. This would not limit those queries.

Reply:
If I make this change, the unit tests I created will fail.

Reply:
@spacedmonkey My code suggestion here is no different from what you have in term of logic flow, the only change is that with my suggestion you don't override the value of `$terms`.

> But what about query where `term_taxonomy_id` are set.

In that case `$args['include']` will never be set, so that should be all good. `$args['include']` is only set if the `default` clause in the `switch` above is made.

Reply:
> In that case $args['include'] will never be set, so that should be all good. $args['include'] is only set if the default clause in the switch above is made.

That is not the desired behaviour here. If you are requesting 5 term_taxonomy_id / name / slug, you should still set the limit. This change would make it so that ONLY term ids are limited, which is not what I am going for here.

Reply:
What is the issue with overriding the terms variable here? Why are people against this? The variable isn't used again.

Reply:
Seems like this change isn't even needed.

Reply:
Ah I see what you mean now, thanks for clarifying!

Reply:
Thanks for switching to `$args['include'] = wp_parse_id_list( $terms );`. 

I see where we (Felix and I) were confused, thanks for your patience and explanation.